File: javascript/lasagna-master/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

/**
 * Implement the functions needed to solve the exercise here.
 * Do not forget to export them so they are available for the
 * tests. Here an example of the syntax as reminder:
 *
 * export function yourFunction(...) {
 *   ...
 * }
 */

function cookingStatus(timer) {
    if (timer === undefined) {
        return 'You forgot to set the timer.'
    }
    return timer === 0 ? 'Lasagna is done.' : 'Not done, please wait.'
}
console.log(cookingStatus(12))
console.log(cookingStatus())
console.log(cookingStatus(0))

function preparationTime(layers, avg = 2) {
    return layers.length * avg
}
const layers = ['sauce', 'noodles', 'sauce', 'meat', 'mozzarella', 'noodles'];
console.log(preparationTime(layers, 3))
console.log(preparationTime(layers))

// function quantities1(layers) {
//     return Object.fromEntries(Object.entries(layers).map(([key, val]) => key === 'noodles' ? [key, val*50]: key === 'sauce' ? [key, val*0.2]: [key, val]))
// }

function quantities(layers) {
    return {
        noodles: layers.filter(layer => layer === 'noodles').length * 50,
        sauce: layers.filter(layer => layer === 'sauce').length * 0.2,
    }
}
console.log("quantities", quantities(layers))

function addSecretIngredient(friendsList, myList) {
    myList.push(friendsList[friendsList.length - 1])
}
const friendsList = ['noodles', 'sauce', 'mozzarella', 'kampot pepper'];
const myList = ['noodles', 'meat', 'sauce', 'mozzarella'];
addSecretIngredient(friendsList, myList)
console.log(myList)

function scaleRecipe(recipe, scale) {
    return Object.fromEntries(Object.entries(recipe).map(([key, value])=> [key, value*scale/2]))
}
const recipe = {
    noodles: 200,
    sauce: 0.5,
    mozzarella: 1,
    meat: 100,
};
console.log(scaleRecipe(recipe, 2))

export {cookingStatus, preparationTime, addSecretIngredient, scaleRecipe, quantities}

File: go/lasagna-master/lasagna_master.go
package lasagna

// TODO: define the 'PreparationTime()' function
func PreparationTime(layers []string, avgTime int) int {
	if avgTime == 0 {
		avgTime = 2
	}
	return len(layers) * avgTime
}

// TODO: define the 'Quantities()' function
func Quantities(layers []string) (noodles int, sauce float64) {
	for _, v := range layers {
		if v == "sauce" {
			sauce += 0.2
		}
		if v == "noodles" {
			noodles += 50
		}
	}
	return
}

// TODO: define the 'AddSecretIngredient()' function
func AddSecretIngredient(friendsList []string, myList []string) {
	myList[len(myList)-1] = friendsList[len(friendsList)-1]
}

// TODO: define the 'ScaleRecipe()' function
func ScaleRecipe(q []float64, n int) []float64 {
	r := make([]float64, len(q))
	for i, v := range q {
		r[i] = v * float64(n) / 2
	}
	return r
}

// Your first steps could be to read through the tasks, and create
// these functions with their correct parameter lists and return types.
// The function body only needs to contain `panic("")`.
//
// This will make the tests compile, but they will fail.
// You can then implement the function logic one by one and see
// an increasing number of tests passing as you implement more
// functionality.
